python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,259 @@
+    """ID of the current table snapshot."""
+
+    snapshots: list = Field(default_factory=list)
+    """A list of valid snapshots. Valid snapshots are snapshots for which
+    all data files exist in the file system. A data file must not be
+    deleted from the file system until the last snapshot in which it was
+    listed is garbage collected."""
+
+    snapshot_log: List[Dict[str, Any]] = Field(alias="snapshot-log", default_factory=list)
+    """A list (optional) of timestamp and snapshot ID pairs that encodes
+    changes to the current snapshot for the table. Each time the
+    current-snapshot-id is changed, a new entry should be added with the
+    last-updated-ms and the new current-snapshot-id. When snapshots are
+    expired from the list of valid snapshots, all entries before a snapshot
+    that has expired should be removed."""
+
+    metadata_log: List[Dict[str, Any]] = Field(alias="metadata-log", default_factory=list)
+    """A list (optional) of timestamp and metadata file location pairs that
+    encodes changes to the previous metadata files for the table. Each time
+    a new metadata file is created, a new entry of the previous metadata
+    file location should be added to the list. Tables can be configured to
+    remove oldest metadata log entries and keep a fixed-size log of the most
+    recent entries after a commit."""
+
+    sort_orders: List[Dict[str, Any]] = Field(alias="sort-orders", default_factory=list)
+    """A list of sort orders, stored as full sort order objects."""
+
+    default_sort_order_id: int = Field(alias="default-sort-order-id")
+    """Default sort order id of the table. Note that this could be used by
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+
+class TableMetadataV1(TableMetadataCommonFields, IcebergBaseModel):
+
+    # When we read a V1 format-version, we'll make sure to populate the fields
+    # for V2 as well. This makes it easier downstream because we can just
+    # assume that everything is a TableMetadataV2.
+    # When writing, we should stick to the same version that it was,
+    # because bumping the version should be an explicit operation that is up
+    # to the owner of the table.
+
+    @root_validator(pre=True)
+    def set_schema_id(cls, data: Dict[str, Any]):
+        # Set some sensible defaults for V1, so we comply with the schema
+        # this is in pre=True, meaning that this will be done before validation
+        # we don't want to make them optional, since we do require them for V2
+        data["schema"]["schema-id"] = DEFAULT_SCHEMA_ID
+        data["default-spec-id"] = INITIAL_SPEC_ID
+        data["last-partition-id"] = max(spec["field-id"] for spec in data["partition-spec"])
+        data["default-sort-order-id"] = 0
+        return data
+
+    @root_validator()
+    def migrate_schema(cls, data: Dict[str, Any]):
+        # Migrate schemas
+        schema = data["schema_"]
```

Comment:
Don't we use the `schema` alias for serialization? Or is that translation done in the JSON to dict mapping, so we will have `schema_` in the dict?

Reply:
So, this root_validator is after the aliases have applied (otherwise it would be `@root_validator(pre=True)`. This means that the reserved Pydantic `.schema` is already translated to `.schema_`. When we serilaize it, we go back to `schema`:
```json
{
	"location": "s3://bucket/test/location",
	"last-updated-ms": 1602638573874,
	"last-column-id": 3,
	"schemas": [{
		"fields": [{
			"id": 1,
			"name": "x",
			"type": "long",
			"required": true
		}, {
			"id": 2,
			"name": "y",
			"type": "long",
			"required": true,
			"doc": "comment"
		}, {
			"id": 3,
			"name": "z",
			"type": "long",
			"required": true
		}],
		"schema-id": 0,
		"identifier-field-ids": []
	}],
	"current-schema-id": 0,
	"partition-specs": [{
		"name": "x",
		"transform": "identity",
		"source-id": 1,
		"field-id": 1000,
		"spec-id": 0
	}],
	"default-spec-id": 0,
	"last-partition-id": 1000,
	"properties": {},
	"current-snapshot-id": -1,
	"snapshots": [{
		"snapshot-id": 1925,
		"timestamp-ms": 1602638573822
	}],
	"snapshot-log": [],
	"metadata-log": [],
	"sort-orders": [],
	"default-sort-order-id": 0,
	"table-uuid": "d20125c8-7284-442c-9aea-15fee620737c",
	"format-version": 1,
	"schema": {
		"fields": [{
			"id": 1,
			"name": "x",
			"type": "long",
			"required": true
		}, {
			"id": 2,
			"name": "y",
			"type": "long",
			"required": true,
			"doc": "comment"
		}, {
			"id": 3,
			"name": "z",
			"type": "long",
			"required": true
		}],
		"schema-id": 0,
		"identifier-field-ids": []
	},
	"partition-spec": [{
		"name": "x",
		"transform": "identity",
		"source-id": 1,
		"field-id": 1000
	}]
}
```